src/mapping/query/function/starts_with.rs

```diff
@@ -0,0 +1,152 @@
+use super::prelude::*;
+
+#[derive(Debug)]
+pub(in crate::mapping) struct StartsWithFn {
+    query: Box<dyn Function>,
+    substring: Box<dyn Function>,
+    case_sensitive: Option<Box<dyn Function>>,
+}
+
+impl StartsWithFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(
+        query: Box<dyn Function>,
+        substring: &str,
+        case_sensitive: bool,
+    ) -> Self {
+        let substring = Box::new(Literal::from(Value::from(substring)));
+        let case_sensitive = Some(Box::new(Literal::from(Value::from(case_sensitive))) as _);
+
+        Self {
+            query,
+            substring,
+            case_sensitive,
+        }
+    }
+}
+
+impl Function for StartsWithFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let substring = {
+            let bytes = required!(ctx, self.substring, Value::Bytes(v) => v);
+            String::from_utf8_lossy(&bytes).into_owned()
+        };
+
+        let value = {
+            let bytes = required!(ctx, self.query, Value::Bytes(v) => v);
+            String::from_utf8_lossy(&bytes).into_owned()
+        };
+
+        let starts_with = value.starts_with(&substring)
+            || optional!(ctx, self.case_sensitive, Value::Boolean(b) => b)
+                .iter()
```

Comment:
Nit: it does look like there is a `filter()` directly on `Option` that might save you from the intermediate iterator.

Reply:
Oh nice. Unfortunately, the ensuing `any` call does need an iterator, so it is needed here.

Reply:
:+1: it seems like we could use `and` there instead? Not a big deal though, mostly just an academic exercise.

Reply:
Well, academic exercises are why I enjoy programming so much! :-)

I think `and` returns an `Option` and I'm trying to get a `bool` out. This could be fixed with an `unwrap_or`. This seems to work:

```rust
let starts_with = value.starts_with(&substring)
            || optional!(ctx, self.case_sensitive, Value::Boolean(b) => b)
                .filter(|&case_sensitive| !case_sensitive)
                .and_then(|_| {
                    Some(value.to_lowercase().starts_with(&substring.to_lowercase()))
                })
            .unwrap_or(false);
```

I think that is nicer. Maybe I shouldn't have been so hasty to merge. I was largely copying the already existing `contains` function..

Reply:
Actually, that `and_then` is really just `map`:

```rust
let starts_with = value.starts_with(&substring)
            || optional!(ctx, self.case_sensitive, Value::Boolean(b) => b)
                .filter(|&case_sensitive| !case_sensitive)
                .map(|_| {
                    value.to_lowercase().starts_with(&substring.to_lowercase())
                })
            .unwrap_or(false);
```

Reply:
:smile: I like that last one too.

This might just be my unfamiliarity with pipeline patterns, but I'd probably find:

```rust
let starts_with = value.starts_with(&substring)
            || (
                 !optional!(ctx, self.case_sensitive, Value::Boolean(b) => b) .unwrap_or(false) && value.to_lowercase().starts_with(&substring.to_lowercase())
               )
```

more readable.

That more clearly says to me that I should get the optional value, or default it to false if unset, and then do the subsequent check.  Albeit, the boolean logic is a bit dense.

Reply:
Yeah, that boolean logic hurts my head. Especially the double negative of `!(...).unwrap_or(false)`. I think really, there's no reason why this has to be a one liner. Splitting it up with a few `if` statements might make it clearer.

Reply:
Agreed :+1: